IBM-Swift/Kitura#883 Rewrite cookie parsing implementation

na-gupta opened the pull request:
## Description
- Rewrite Kitura cookie parsing implementation
- Cookie delimiter should be semicolon followed by zero or more whitespace characters
- Use first equals in cookie as name/value delimiter, remaining are treated as part of cookie value
- If cookie value is quoted(starts and ends with `"`), remove quotes
- Ensure unicode characters in cookies values are supported
- Add tests for all these conditions
- Cleanup: remove NSString casts that are no longer needed

## Motivation and Context
Rewrite Kitura cookie parsing implementation to better match standards
Addresses IBM-Swift/Kitura#883 among other things

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [ ] If applicable, I have updated the documentation accordingly.
- [ ] If applicable, I have added tests to cover my changes.

ianpartridge commented on Tests/KituraTests/TestCookies.swift:
```diff
@@ -61,25 +52,38 @@ class TestCookies: XCTestCase {
     let router = TestCookies.setupRouter()
 
-    func testCookieToServer() {
+    func testCookieToServerWithColonSeparator() {
```
Shouldn't this be `testCookieToServerWithSemiColonSeparator()`?

ianpartridge replied:
Thanks. It obviously applies to the other tests too.

ianpartridge commented on Sources/Kitura/RouterRequest.swift:
```diff
@@ -234,42 +225,46 @@ public class RouterRequest {
+            return try RegularExpressionType(pattern: "[^;\\s][^;]*", options: [])
+        } catch { // should never throw here, famous last words
+            Log.error("Error creating cookie separator regex: \(error)")
+            exit(1)
```
Surely this is bad...

na-gupta replied:
I've addressed the other comments, but not sure if there's a better pattern/way to achieve this one?
The pattern literal we're using is pretty straightforward, so any changes to RegularExpression to break this would have to be really bad.

ianpartridge replied:
Crashing Kitura cannot be right...  Also, why do you create a new `RegularExpressionType` every time this computed property is accessed? Compiling the regex is likely to have a non-trivial cost so should only be done once.

na-gupta replied:
Not sure what you mean, I'm only creating the regex once. It is not computed each time.
`private static var separator: RegularExpressionType = {...}()`

na-gupta replied:
I can make the separator regex optional, and let the cooking parsing return an empty dictionary in the unlikely case where it fails. Does that sound like a better approach?

ianpartridge replied:
OK, maybe I'm going mad here, but just because it's `static` doesn't mean it's only computed once. Every time the property is accessed it is recomputed.

Try running this:

```swift
class Test {
    static var aString: String {
        print("test")
        return "string"
    }
}

print(Test.aString)
print(Test.aString)
```

It prints:

```
test
string
test
string
```

ianpartridge replied:
Yes, the optional idea sounds better to me. We need a solution that doesn't bring down the server.

na-gupta replied:
I'm not using a computed value. The code I use is initializing it from a closure return just as if I did:
`private static var separator: RegularExpressionType = createRegex()`

It's similar to:
```
class Test {
    static var aString: String = {
        print("test")
        return "string"
    }()
}

print(Test.aString)
print(Test.aString)
```
For which the output is:
```
test
string
string
```

ianpartridge replied:
!! I totally missed the closure execution. Apologies.

ianpartridge commented on Sources/Kitura/RouterRequest.swift:
```diff
@@ -234,42 +225,46 @@ public class RouterRequest {
-
-                cookies[cookieNameValueParts[0]] = theCookie
-            }
+    private static func getCookie(cookieString: String) -> HTTPCookie? {
```
I think including the type name in the variable name is not best practice.

na-gupta replied:
Ok, will change it

ianpartridge commented on Sources/Kitura/RouterRequest.swift:
```diff
@@ -125,15 +121,10 @@ public class RouterRequest {
     /// Parsed Cookies, used to do a lazy parsing of the appropriate headers.
-    private lazy var _cookies: Cookies = { [unowned self] in
-        return Cookies(headers: self.serverRequest.headers)
+    public lazy var cookies: [String: HTTPCookie] = { [unowned self] in
```
Is there a reason why the private `_cookies` has gone?

ianpartridge replied:
OK, I don't mind removing it, @shmuelk may have had a reason the code was structured this way though.

ianpartridge commented:
I've reviewed this but I would like @shmuelk to take a good look at the changes to cookie handling in case there's anything I've missed.

na-gupta commented:
Remove regex use per discussion with @shmuelk 
Also added support to unquote cookie values as per spec and added tests for it
@shmuelk @ianpartridge Could you please review this again.

ianpartridge commented on Sources/Kitura/RouterRequest.swift:
```diff
@@ -234,42 +225,41 @@ public class RouterRequest {
+
+        for cookieHeader in cookieHeaders {
+            for cookie in cookieHeader.components(separatedBy: ";") {
+                let trimmedCookie = cookie.trimmingCharacters(in: .whitespaces)
```
We didn't trim this before, did we? Is this specified in an RFC or something?

na-gupta replied:
The standard for the separator is semicolons with zero or more whitespace. Shmuel suggested removing the regex, splitting by semicolons and stripping out leading/trailing whitespace, which achieves the goal without using regular expressions.

ianpartridge commented:
This looks good to me now. @shmuelk merge if you're happy?

shmuelk commented:
Sorry.

shmuelk commented:
Looks ok. I'll merge after CI completes.
